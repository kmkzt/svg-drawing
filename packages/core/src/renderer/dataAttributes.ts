import type {
  SelectEventObject,
  SelectEventType,
  ElementKey,
  VertexType,
} from '../types'

const dataEditType = 'data-edit-type'
const dataElementKey = 'data-element-key'
const dataCommandIndex = 'data-command-index'
const dataPointIndex = 'data-point-index'
const dataVertexType = 'data-vertex-type'

export const dataPathAttributes = (elementKey: ElementKey) =>
  ({
    [dataEditType]: 'path',
    [dataElementKey]: elementKey,
  } as const)

export const dataPathCommandAttributes = ({
  elementKey,
  commandIndex,
}: {
  elementKey: ElementKey
  commandIndex: number
}) =>
  ({
    [dataEditType]: 'path/command',
    [dataElementKey]: elementKey,
    [dataCommandIndex]: `${commandIndex}`,
  } as const)

export const dataPathAnchorPointAttributes = ({
  elementKey,
  commandIndex,
  pointIndex,
}: {
  elementKey: ElementKey
  pointIndex: number
  commandIndex: number
}) =>
  ({
    [dataEditType]: 'path/point',
    [dataElementKey]: elementKey,
    [dataCommandIndex]: `${commandIndex}`,
    [dataPointIndex]: `${pointIndex}`,
  } as const)

export const dataBoundingBoxAttributes = {
  [dataEditType]: 'bounding-box',
} as const

export const dataBoundingBoxVertexAttributes = (vertexType: VertexType) =>
  ({
    [dataEditType]: 'bounding-box/vertex',
    [dataVertexType]: vertexType,
  } as const)

export const dataFrameAttributes = {
  [dataEditType]: 'frame',
} as const

export const getSelectEvent = (
  el: HTMLElement,
  multiple = false
): SelectEventObject | null => {
  const type = el.getAttribute(dataEditType) as SelectEventType

  switch (type) {
    case 'path': {
      const key = el.getAttribute(dataElementKey)

      return key
        ? {
            type,
            key,
            multiple,
          }
        : null
    }

    case 'path/command': {
      const key = el.getAttribute(dataElementKey)
      const commandIndex = el.getAttribute(dataCommandIndex)

      return key && commandIndex !== null
        ? {
            type,
            key,
            index: {
              command: +commandIndex,
            },
            multiple,
          }
        : null
    }

    case 'path/point': {
      const key = el.getAttribute(dataElementKey)
      const commandIndex = el.getAttribute(dataCommandIndex)
      const pointIndex = el.getAttribute(dataPointIndex)

      return key && commandIndex !== null && pointIndex !== null
        ? {
            type,
            key,
            index: {
              command: +commandIndex,
              point: +pointIndex,
            },
            multiple,
          }
        : null
    }

    case 'bounding-box': {
      return { type }
    }
    case 'bounding-box/vertex': {
      const vertexType = el.getAttribute(dataVertexType) as VertexType

      return vertexType
        ? {
            type,
            vertexType,
          }
        : null
    }

    case 'frame': {
      return { type }
    }

    default: {
      return null
    }
  }
}
