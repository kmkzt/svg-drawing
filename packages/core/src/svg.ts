import { roundUp } from './shared/roundUp'
import { camel2kebab } from './shared/camel2kebab'
import { kebab2camel } from './shared/kebab2camel'
import { svg2base64 } from './shared/svg2base64'
import {
  createSvgElement,
  createSvgChildElement,
} from './shared/createSvgElement'
import { download } from './shared/download'

const isNaN = (num: number) => num !== num

export class Point {
  public x: number
  public y: number
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  public toVector(): Vector {
    const value = Math.sqrt(Math.pow(this.x, 2.0) + Math.pow(this.y, 2.0))
    const angle = Math.atan2(this.y, this.x)
    return new Vector(value, angle)
  }

  public scale(r: number): Point {
    return new Point(this.x * r, this.y * r)
  }

  public add(p: Point): Point {
    return new Point(this.x + p.x, this.y + p.y)
  }

  public sub(p: Point): Point {
    return new Point(this.x - p.x, this.y - p.y)
  }

  public eql(p: Point): boolean {
    return this.x === p.x && this.y === p.y
  }

  public clone(): Point {
    return new Point(this.x, this.y)
  }
}

export const COMMAND_TYPE = {
  MOVE: 'M', // M 0 0
  MOVE_RELATIVE: 'm', // m 0 0
  LINE: 'L', // L 1 1
  LINE_RELATIVE: 'l', // l 1 1
  CURVE: 'C', // C 1 1 2 2 3 3
  CURVE_RELATIVE: 'c', // c 1 1 2 2 3 3
  CLOSE: 'Z', // Z, z
  HORIZONTAL: 'H', // H 10
  HORIZONTAL_RELATIVE: 'h', // h 10
  VERTICAL: 'V', // V 20
  VERTICAL_RELATIVE: 'v', // v 20
  ARC_CURVE: 'A', // A 6 4 10 0 1 14 10
  ARC_CURVE_RELATIVE: 'a', // A 6 4 10 0 1 14 10
  QUADRATIC_CURVE: 'Q', // Q 10 60 10 30
  QUADRATIC_CURVE_RELATIVE: 'q', // q 10 60 10 30
} as const

type COMMAND = typeof COMMAND_TYPE[keyof typeof COMMAND_TYPE]
// TODO: compatible COMMAND_TYPE
export class Command {
  public type: COMMAND
  public value: number[]
  // TODO: Convert data format to number array.
  constructor(type: COMMAND, value: number[] = []) {
    this.value = value
    this.type = type
  }

  public set cr(po: Point | undefined) {
    if (!po) return
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return
    }
    this.value.splice(2, 1, po.x)
    this.value.splice(3, 1, po.y)
  }
  public get cr(): Point | undefined {
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return undefined
    }
    const [x, y] = this.value.slice(2, 4)
    return new Point(x, y)
  }
  public set cl(po: Point | undefined) {
    if (!po) return
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return
    }
    this.value.splice(0, 1, po.x)
    this.value.splice(1, 1, po.y)
  }
  public get cl(): Point | undefined {
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return undefined
    }
    const [x, y] = this.value.slice(0, 2)
    return new Point(x, y)
  }

  public set point(po: Point | undefined) {
    if (!po) return
    this.value.splice(this.value.length - 2, 1, po.x)
    this.value.splice(this.value.length - 1, 1, po.y)
  }
  public get point(): Point | undefined {
    const xy = this.value.slice(this.value.length - 2)
    return xy.length === 2 ? new Point(xy[0], xy[1]) : undefined
  }

  public toString(): string {
    if (this.type === COMMAND_TYPE.CLOSE) return COMMAND_TYPE.CLOSE
    return `${this.type} ${this.value.map((v) => roundUp(v)).join(' ')}`
  }

  public scale(r: number): Command {
    const upd = new Command(
      this.type,
      this.value.map((p) => p * r)
    )
    return upd
  }

  public clone(): Command {
    return new Command(this.type, this.value.slice())
  }
}

export class Vector {
  public angle: number
  public value: number
  constructor(v: number, a: number) {
    this.value = v
    this.angle = a
  }

  public toPoint(): Point {
    const x = Math.cos(this.angle) * this.value
    const y = Math.sin(this.angle) * this.value
    return new Point(x, y)
  }

  public scale(r: number): Vector {
    return new Vector(this.value * r, this.angle)
  }
}

// TODO: improve key types
export interface PathObject {
  [key: string]: string | undefined
}

/**
 * TODO: refactor command.
 * The following commands are not supported. Cannot support commands that use `M` or` z` more than once
 * `M 0 0 L 1 1 Z M 1 1 L 2 2 Z`
 */
export class Path {
  public attrs: PathObject
  public commands: Command[]

  constructor({ d, ...attrs }: PathObject = {}) {
    this.attrs = attrs
    this.commands = []
    if (d) this.parseCommandString(d)
  }

  public scale(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scale(r))
    this.attrs.strokeWidth = String(r * +(this.attrs.strokeWidth || 0))
    return this
  }

  public addCommand(param: Command | Command[]): this {
    if (Array.isArray(param)) {
      this.commands.push(...param)
    } else {
      this.commands.push(param)
    }
    return this
  }
  public getCommandString(): string {
    if (this.commands.length === 0) return ''
    return this.commands
      .map((com: Command, _i: number) => com.toString())
      .join(' ')
      .trim()
  }

  // TODO: Valid Command type
  public parseCommandString(d: string): void {
    this.commands = []
    let type: COMMAND | null = null
    let value: number[] = []
    const c = d.split(' ')
    const checkType = (c: any): COMMAND | null =>
      Object.values(COMMAND_TYPE).includes(c) ? c : null
    for (let i = 0; i < c.length; i += 1) {
      const t = checkType(c[i])
      // COMMAND Parse
      if (t) {
        if (!type) {
          type = t
          continue
        }
        this.commands.push(new Command(type, value))
        type = t
        value = []
        continue
      }
      if (isNaN(+c[i])) {
        return
      }
      value.push(+c[i])
    }

    if (type !== null) {
      this.commands.push(new Command(type, value))
    }
  }

  public parsePathElement(pEl: SVGPathElement): this {
    for (let i = 0; i < pEl.attributes.length; i += 1) {
      const attr: Attr | null = pEl.attributes.item(i)
      if (!attr || !attr.value) continue
      if (attr.name === 'd') {
        this.parseCommandString(attr.value)
        continue
      }
      this.attrs = {
        ...this.attrs,
        [kebab2camel(attr.name)]: attr.value,
      }
    }
    return this
  }

  public toJson(): PathObject {
    return {
      ...this.attrs,
      d: this.getCommandString(),
    }
  }
  public toElement(): SVGElement {
    const attrs = Object.entries(this.attrs).reduce(
      (acc, [key, val], _i) =>
        val
          ? {
              ...acc,
              [camel2kebab(key)]: val,
            }
          : acc,
      {}
    )
    return createSvgChildElement('path', {
      ...attrs,
      d: this.getCommandString(),
    })
  }

  public clone(): Path {
    const path = new Path(this.attrs)
    this.commands.map((c) => {
      path.commands.push(c.clone())
    })
    return path
  }
}

export interface SvgOption {
  width: number
  height: number
  background?: string
}

export interface SvgObject {
  width: number
  height: number
  background?: string
  paths: PathObject[]
}
export class Svg {
  public paths: Path[]
  public width: number
  public height: number
  public background?: string

  constructor({ width, height, background }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
    this.background = background
  }

  public scalePath(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public addPath(pa: Path | Path[]): this {
    if (Array.isArray(pa)) {
      this.paths.push(...pa)
    } else {
      this.paths.push(pa)
    }
    return this
  }
  public clonePaths(): Path[] {
    return this.paths.map((p) => p.clone())
  }

  public updatePath(pa: Path, i?: number): this {
    const updateIndex = i || this.paths.length - 1
    if (updateIndex < 0) this.paths.push(pa)
    this.paths[updateIndex] = pa
    return this
  }

  public toJson(): SvgObject {
    return {
      width: this.width,
      height: this.height,
      background: this.background,
      paths: this.paths.map((p) => p.toJson()),
    }
  }

  public copy(svg: any extends Svg ? Svg : never): this {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scalePath(this.width / svg.width)
    }
    return this
  }

  public toElement(): SVGSVGElement {
    const size = { width: String(this.width), height: String(this.height) }
    const bgEl = this.background
      ? [createSvgChildElement('rect', { ...size, fill: this.background })]
      : []

    return createSvgElement(
      { width: String(this.width), height: String(this.height) },
      bgEl.concat(this.paths.map((p) => p.toElement()))
    )
  }
  public toBase64(): string {
    return svg2base64(this.toElement().outerHTML)
  }

  public parseSVGString(svgStr: string): this {
    const svgEl: SVGSVGElement | null = new DOMParser()
      .parseFromString(svgStr, 'image/svg+xml')
      .querySelector('svg')
    if (!svgEl) {
      this.paths = []
      return this
    }
    return this.parseSVGElement(svgEl)
  }

  public parseSVGElement(svgEl: SVGSVGElement): this {
    const update: Path[] = []
    svgEl.querySelectorAll('path').forEach((pEl) => {
      const pa = new Path().parsePathElement(pEl)
      if (pa.commands.length !== 0) {
        update.push(pa)
      }
    })
    this.paths = update
    const width = Number(svgEl.getAttribute('width'))
    if (width && this.width) {
      this.scalePath(this.width / width)
    }
    return this
  }

  // TODO: Add filename config
  public download(
    ext: 'svg' | 'png' | 'jpg' = 'svg',
    cb: typeof download = download
  ): void {
    if (ext === 'svg') {
      cb({
        data: this.toBase64(),
        extension: 'svg',
      })
      return
    }

    const img: any = new Image()
    const renderCanvas = () => {
      const canvas = document.createElement('canvas')
      canvas.setAttribute('width', String(this.width))
      canvas.setAttribute('height', String(this.height))
      const ctx = canvas.getContext('2d')
      if (!ctx) return
      if (this.background || ext === 'jpg') {
        ctx.fillStyle = this.background || '#fff'
        ctx.fillRect(0, 0, this.width, this.height)
      }
      ctx.drawImage(img, 0, 0)
      if (ext === 'png') {
        cb({ data: canvas.toDataURL('image/png'), extension: 'png' })
      } else {
        cb({ data: canvas.toDataURL('image/jpeg'), extension: 'jpg' })
      }
    }
    img.addEventListener('load', renderCanvas, false)
    img.src = this.toBase64()
  }
}
