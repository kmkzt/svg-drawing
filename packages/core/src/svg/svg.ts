import type { SvgOption, PathClass, SvgClass } from '../types'

export class Svg implements SvgClass {
  public paths: PathClass[]
  public width: number
  public height: number
  public background?: string

  constructor({ width, height, background }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
    this.background = background
  }

  /** @todo Implement to be resized according to height */
  public resize({ width, height }: { width: number; height: number }) {
    this.scale(width / this.width)
    this.width = width
    this.height = height
  }

  private scale(r: number) {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public addPath(pa: PathClass | PathClass[]) {
    if (Array.isArray(pa)) {
      this.paths.push(...pa)
    } else {
      this.paths.push(pa)
    }
    return this
  }

  public getPath(key: string): PathClass | undefined {
    return this.paths.find((p) => p.key === key)
  }

  public updatePath(path: PathClass) {
    const index = this.paths.findIndex((p) => p.key === path.key)
    if (index !== -1) {
      this.paths[index] = path
    }
    return this
  }

  public deletePath(deletePath: PathClass) {
    this.paths = this.paths.filter((path) => path.key !== deletePath.key)
    return this
  }

  public clonePaths() {
    return this.paths.map((p) => p.clone())
  }

  public toJson() {
    return {
      width: this.width,
      height: this.height,
      background: this.background,
      paths: this.paths.map((p) => p.toJson()),
    }
  }

  /**
   * Copy resized paths.
   *
   * @example Const drawSvg = new Svg() const animateSvg = new Svg().copy(drawSvg)
   */
  public copy(svg: SvgClass) {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scale(this.width / svg.width)
    }
    return this
  }

  public clone() {
    const svg = new Svg({
      width: this.width,
      height: this.height,
      background: this.background,
    })
    svg.addPath(this.clonePaths())
    return svg
  }
}
