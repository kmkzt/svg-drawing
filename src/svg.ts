import { roundUp } from './utils/roundUp'
import { camel2kebab } from './utils/camel2kebab'
import { kebab2camel } from './utils/kebab2camel'
import { svg2base64 } from './utils/svg2base64'
import {
  createSvgElement,
  createSvgChildElement
} from './utils/createSvgElement'

const isNaN = (num: number) => num !== num

export class Point {
  public x: number
  public y: number
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  public toVector(): Vector {
    const value = Math.sqrt(Math.pow(this.x, 2.0) + Math.pow(this.y, 2.0))
    const angle = Math.atan2(this.y, this.x)
    return new Vector(value, angle)
  }

  public scale(r: number): Point {
    return new Point(this.x * r, this.y * r)
  }

  public add(p: Point): Point {
    return new Point(this.x + p.x, this.y + p.y)
  }

  public sub(p: Point): Point {
    return new Point(this.x - p.x, this.y - p.y)
  }

  public eql(p: Point): boolean {
    return this.x === p.x && this.y === p.y
  }

  public clone(): Point {
    return new Point(this.x, this.y)
  }
}

export const COMMAND_TYPE = {
  MOVE: 'M', // M 0 0
  MOVE_RELATIVE: 'm', // m 0 0
  LINE: 'L', // L 1 1
  LINE_RELATIVE: 'l', // l 1 1
  CURVE: 'C', // C 1 1 2 2 3 3
  CURVE_RELATIVE: 'c', // c 1 1 2 2 3 3
  CLOSE: 'Z', // Z, z
  HORIZONTAL: 'H', // H 10
  HORIZONTAL_RELATIVE: 'h', // h 10
  VERTICAL: 'V', // V 20
  VERTICAL_RELATIVE: 'v', // v 20
  ARC_CURVE: 'A', // A 6 4 10 0 1 14 10
  ARC_CURVE_RELATIVE: 'a', // A 6 4 10 0 1 14 10
  QUADRATIC_CURVE: 'Q', // Q 10 60 10 30
  QUADRATIC_CURVE_RELATIVE: 'q' // q 10 60 10 30
} as const

type COMMAND = typeof COMMAND_TYPE[keyof typeof COMMAND_TYPE]
// TODO: compatible COMMAND_TYPE
export class Command {
  public type: COMMAND
  public value: number[]
  // TODO: Convert data format to number array.
  constructor(type: COMMAND, value: number[] = []) {
    this.value = value
    this.type = type
  }

  public set cr(po: Point | undefined) {
    if (!po) return
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return
    }
    this.value.splice(2, 1, po.x)
    this.value.splice(3, 1, po.y)
  }
  public get cr(): Point | undefined {
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return undefined
    }
    const [x, y] = this.value.slice(2, 4)
    return new Point(x, y)
  }
  public set cl(po: Point | undefined) {
    if (!po) return
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return
    }
    this.value.splice(0, 1, po.x)
    this.value.splice(1, 1, po.y)
  }
  public get cl(): Point | undefined {
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return undefined
    }
    const [x, y] = this.value.slice(0, 2)
    return new Point(x, y)
  }

  public set point(po: Point | undefined) {
    if (!po) return
    this.value.splice(this.value.length - 2, 1, po.x)
    this.value.splice(this.value.length - 1, 1, po.y)
  }
  public get point(): Point | undefined {
    const xy = this.value.slice(this.value.length - 2)
    return xy.length === 2 ? new Point(xy[0], xy[1]) : undefined
  }

  public toString(): string {
    return `${this.type} ${this.value.map(v => roundUp(v)).join(' ')}`
  }

  public scale(r: number): Command {
    const upd = new Command(
      this.type,
      this.value.map(p => p * r)
    )
    return upd
  }

  public clone(): Command {
    const copy = new Command(this.type, this.value.slice())
    return copy
  }
}

export class Vector {
  public angle: number
  public value: number
  constructor(v: number, a: number) {
    this.value = v
    this.angle = a
  }

  public toPoint(): Point {
    const x = Math.cos(this.angle) * this.value
    const y = Math.sin(this.angle) * this.value
    return new Point(x, y)
  }

  public scale(r: number): Vector {
    return new Vector(this.value * r, this.angle)
  }
}

export interface PathOption {
  strokeWidth?: number
  attrs?: Attrs
  stroke?: string
  fill?: string
}

interface Attrs {
  [key: string]: string
}
export interface PathObject {
  attrs: Attrs
  d: string
}

const defaultAttrs: Attrs = {
  strokeLinecap: 'round', // 'mitter'
  strokeLinejoin: 'round' // 'square'
}
/**
 * TODO: refactor command.
 * The following commands are not supported. Cannot support commands that use `M` or` z` more than once
 * `M 0 0 L 1 1 Z M 1 1 L 2 2 Z`
 */
export class Path {
  public strokeWidth: number
  public attrs: { [key: string]: string }
  public stroke: string
  public fill: string
  public commands: Command[]

  constructor({ strokeWidth, fill, stroke, attrs }: PathOption = {}) {
    this.strokeWidth = strokeWidth ?? 1
    this.stroke = stroke ?? '#000'
    this.fill = fill ?? 'none'
    this.attrs = {
      ...defaultAttrs,
      ...attrs
    }
    this.commands = []
  }

  public scale(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scale(r))
    this.strokeWidth *= r
    return this
  }

  public addCommand(param: Command): this {
    this.commands.push(param)
    return this
  }
  public getCommandString(): string {
    if (this.commands.length === 0) return ''
    return this.commands
      .map((com: Command, _i: number) => com.toString())
      .join(' ')
      .trim()
  }

  // TODO: Valid Command type
  public parseCommandString(d: string): void {
    this.commands = []
    let type: COMMAND | null = null
    let value: number[] = []
    const c = d.split(' ')
    const checkType = (c: any): COMMAND | null =>
      Object.values(COMMAND_TYPE).includes(c) ? c : null
    for (let i = 0; i < c.length; i += 1) {
      const t = checkType(c[i])
      // COMMAND Parse
      if (t) {
        if (!type) {
          type = t
          continue
        }
        this.commands.push(new Command(type, value))
        type = t
        value = []
        continue
      }
      if (isNaN(+c[i])) {
        return
      }
      value.push(+c[i])
    }

    if (type !== null) {
      this.commands.push(new Command(type, value))
    }
  }

  public parsePathElement(pEl: SVGPathElement): this {
    this.strokeWidth = 1
    this.stroke = '#000'
    this.fill = 'none'
    this.attrs = defaultAttrs
    for (let i = 0; i < pEl.attributes.length; i += 1) {
      const attr: Attr | null = pEl.attributes.item(i)
      if (!attr) continue
      if (attr.name === 'd') {
        this.parseCommandString(attr.value)
        continue
      }
      if (attr.name === 'stroke-width') {
        this.strokeWidth = Number(attr.value)
        continue
      }
      if (attr.name === 'fill') {
        this.fill = attr.value
        continue
      }
      if (attr.name === 'stroke') {
        this.stroke = attr.value
        continue
      }
      this.attrs = {
        ...this.attrs,
        [kebab2camel(attr.name)]: attr.value
      }
    }
    return this
  }

  public toJson(): PathObject {
    return {
      attrs: this.attrs,
      d: this.getCommandString()
    }
  }
  public toElement(): SVGElement {
    return createSvgChildElement('path', {
      fill: this.fill,
      stroke: this.stroke,
      ['stroke-width']: String(this.strokeWidth),
      d: this.getCommandString(),
      ...Object.entries(this.attrs).reduce(
        (acc, [key, val], _i) => ({
          ...acc,
          [camel2kebab(key)]: val
        }),
        {}
      )
    })
  }

  public clone(): Path {
    const path = new Path({
      strokeWidth: this.strokeWidth,
      fill: this.fill,
      stroke: this.stroke,
      attrs: { ...this.attrs }
    })
    this.commands.map(c => {
      path.commands.push(c.clone())
    })
    return path
  }
}

export interface SvgOption {
  width: number
  height: number
}

export interface SvgObject {
  width: number
  height: number
  paths: PathObject[]
}
export class Svg {
  public paths: Path[]
  public width: number
  public height: number
  constructor({ width, height }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
  }

  public scalePath(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public addPath(pa: Path): this {
    this.paths.push(pa)
    return this
  }
  public clonePaths(): Path[] {
    return this.paths.map(p => p.clone())
  }

  public updatePath(pa: Path, i?: number): this {
    const updateIndex = i || this.paths.length - 1
    if (updateIndex < 0) this.paths.push(pa)
    this.paths[updateIndex] = pa
    return this
  }

  public toJson(): SvgObject {
    return {
      width: this.width,
      height: this.height,
      paths: this.paths.map(p => p.toJson())
    }
  }

  public copy(svg: any extends Svg ? Svg : never): this {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scalePath(svg.width / this.width)
    }
    return this
  }

  public toElement(): SVGSVGElement {
    return createSvgElement(
      { width: String(this.width), height: String(this.height) },
      this.paths.map(p => p.toElement())
    )
  }
  public toBase64(): string {
    return svg2base64(this.toElement().outerHTML)
  }

  public parseSVGString(svgStr: string): this {
    const svgEl: SVGSVGElement | null = new DOMParser()
      .parseFromString(svgStr, 'image/svg+xml')
      .querySelector('svg')
    if (!svgEl) {
      this.paths = []
      return this
    }
    return this.parseSVGElement(svgEl)
  }

  public parseSVGElement(svgEl: SVGSVGElement): this {
    const update: Path[] = []
    svgEl.querySelectorAll('path').forEach(pEl => {
      const pa = new Path().parsePathElement(pEl)
      if (pa.commands.length !== 0) {
        update.push(pa)
      }
    })
    this.paths = update
    const width = Number(svgEl.getAttribute('width'))
    if (width && this.width) {
      this.scalePath(this.width / width)
    }
    return this
  }
}
