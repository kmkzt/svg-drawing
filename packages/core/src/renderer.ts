import { Svg, SvgOption } from './svg'

export type RendererOption = Omit<SvgOption, 'width' | 'height'>
export class Renderer {
  public el: HTMLElement
  public top: number
  public left: number
  public svg: Svg
  constructor(el: HTMLElement, opt: RendererOption = {}) {
    const { width, height, left, top } = el.getBoundingClientRect()
    /**
     * Setup parameter
     */
    this.el = el
    this.left = left
    this.top = top
    this.svg = new Svg({ width, height, ...opt })
    el.appendChild(this.svg.toElement())
    this._setupAdjustResize()
  }
  /**
   * render
   * TODO: XSS test
   */
  public update(): void {
    this.el.replaceChild(this.svg.toElement(), this.el.childNodes[0])
  }

  public resizeElement(param?: DOMRect): void {
    const { width, height, left, top } =
      param || this.el.getBoundingClientRect()
    // TODO: Resizing improve
    this.svg.scalePath(width / this.svg.width)
    this.svg.width = width
    this.svg.height = height
    this.left = left
    this.top = top
  }

  private _setupAdjustResize(): void {
    if ((window as any).ResizeObserver) {
      const resizeObserver: any = new (window as any).ResizeObserver(
        (entries: any[]) => {
          this.resizeElement(entries[0].contentRect)
          this.update()
        }
      )
      resizeObserver.observe(this.el)
    } else {
      // TODO: improve
      window.addEventListener('resize', (_ev) => {
        this.resizeElement(this.el.getBoundingClientRect())
        this.update()
      })
    }
  }
}
