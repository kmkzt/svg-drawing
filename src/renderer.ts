import { Svg, Path } from './svg'
import { download } from './utils/download'

export interface RendererOption {
  background?: string
}
export class Renderer extends Svg {
  public el: HTMLElement
  public top: number
  public left: number
  public background: string
  constructor(el: HTMLElement, { background }: RendererOption = {}) {
    const { width, height, left, top } = el.getBoundingClientRect()
    super({ width, height })
    /**
     * Setup parameter
     */
    this.el = el
    this.left = left
    this.top = top
    this.background = background ?? '#fff'
    el.appendChild(this.toElement())
    this._setupAdjustResize()
  }
  /**
   * render
   * TODO: XSS test
   */
  public update(): void {
    this.el.replaceChild(this.toElement(), this.el.childNodes[0])
  }

  public resizeElement(param?: DOMRect): void {
    const { width, height, left, top } =
      param || this.el.getBoundingClientRect()
    // TODO: Resizing improve
    this.scalePath(width / this.width)
    this.width = width
    this.height = height
    this.left = left
    this.top = top
  }

  private _setupAdjustResize(): void {
    if ((window as any).ResizeObserver) {
      const resizeObserver: any = new (window as any).ResizeObserver(
        (entries: any[]) => {
          this.resizeElement(entries[0].contentRect)
          this.update()
        }
      )
      resizeObserver.observe(this.el)
    } else {
      // TODO: improve
      window.addEventListener('resize', (_ev) => {
        this.resizeElement(this.el.getBoundingClientRect())
        this.update()
      })
    }
  }

  // TODO: Add filename config
  public download(
    ext: 'svg' | 'png' | 'jpg' = 'svg',
    cb: typeof download = download
  ): void {
    if (ext === 'svg') {
      cb({
        data: this.toBase64(),
        extension: 'svg',
      })
      return
    }

    const img: any = new Image()
    const renderCanvas = () => {
      const canvas = document.createElement('canvas')
      canvas.setAttribute('width', String(this.width))
      canvas.setAttribute('height', String(this.height))
      const ctx = canvas.getContext('2d')
      if (!ctx) return
      ctx.fillStyle = this.background
      ctx.fillRect(0, 0, this.width, this.height)
      ctx.drawImage(img, 0, 0)
      if (ext === 'png') {
        cb({ data: canvas.toDataURL('image/png'), extension: 'png' })
      } else {
        cb({ data: canvas.toDataURL('image/jpeg'), extension: 'jpg' })
      }
    }
    img.addEventListener('load', renderCanvas, false)
    img.src = this.toBase64()
  }
}
