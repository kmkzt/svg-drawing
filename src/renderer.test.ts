import { Renderer } from './renderer'
import { Path, Point } from './svg'

describe('renderer.ts', () => {
  describe('Renderer', () => {
    let renderer: Renderer
    beforeEach(() => {
      const el = document.createElement('div')
      renderer = new Renderer(el)
        .addPath(
          new Path({ curve: true, close: false })
            .addCommand(new Point(0, 0))
            .addCommand(new Point(1, 1))
            .addCommand(new Point(2, 1))
            .addCommand(new Point(3, 0))
        )
        .addPath(
          new Path({ curve: false, close: true })
            .addCommand(new Point(4, 4))
            .addCommand(new Point(9, 4))
            .addCommand(new Point(9, 8))
            .addCommand(new Point(3, 0))
        )
    })
    it('download svg', done => {
      const testDownload = (param: any): void => {
        expect(param).toMatchSnapshot()
        done()
      }
      renderer.download('svg', testDownload)
    })
    // TODO: Fix download test
    // it('download jpg', done => {
    //   const testDownload = (param: any): void => {
    //     expect(param).toMatchSnapshot()
    //     done()
    //   }
    //   renderer.download('jpg', testDownload)
    // }, 30000)
    // it('download png', done => {
    //   const testDownload = (param: any): void => {
    //     expect(param).toMatchSnapshot()
    //     done()
    //   }
    //   renderer.download('png', testDownload)
    // }, 30000)
  })
})
