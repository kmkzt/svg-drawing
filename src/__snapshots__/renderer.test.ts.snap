// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer.ts Renderer download svg 1`] = `
Object {
  "data": "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMCIgaGVpZ2h0PSIwIiB2ZXJzaW9uPSIxLjEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PHBhdGggc3Ryb2tlLXdpZHRoPSIxIiBzdHJva2U9IiMwMDAiIGZpbGw9Im5vbmUiIHN0cm9rZS1saW5lam9pbj0icm91bmQiIHN0cm9rZS1saW5lY2FwPSJyb3VuZCIgZD0iTSAwIDAgQyAwLjIgMC4yIDAuNiAwLjggMSAxIEMgMS40IDEuMiAxLjYgMS4yIDIgMSBDIDIuNCAwLjggMi44IDAuMiAzIDAiPjwvcGF0aD48cGF0aCBzdHJva2Utd2lkdGg9IjEiIHN0cm9rZT0iIzAwMCIgZmlsbD0ibm9uZSIgc3Ryb2tlLWxpbmVqb2luPSJtaXR0ZXIiIHN0cm9rZS1saW5lY2FwPSJzcXVhcmUiIGQ9Ik0gNCA0IEwgOSA0IEwgOSA4IEwgMyAwIFoiPjwvcGF0aD48L3N2Zz4=",
  "extension": "svg",
}
`;
